src/lava/proc/io/bridge/extractor.py

```diff
@@ -0,0 +1,179 @@
+    receive data in external Python scripts from the ProcessModel of the
+    Process, while the Lava Runtime is running.
+    Internally, this Process builds a channel (extractor_channel, of type
+    PyPyChannel).
+    The src_port of the channel lives in the ProcessModel.
+    The dst_port of the channel lives in the Process.
+
+    In the ProcessModel, data is received from this Process's InPort,
+    and relayed to the extractor_channel.src_port.
+    When the recv_data is called from the external Python script, data is
+    received from the extractor_channel.dst_port.
+
+    For the sending part of the extractor_channel (src_port), the following
+    property can be parametrized:
+        (1) When the channel buffer is full, sending new data will either:
+            (a) Block until free space is available (until the receiving part
+            receives an item).
+            (b) Not block, and the new data will not be sent.
+
+    For the receiving part of the extractor_channel (recv_port), the following
+    properties can be parametrized:
+        (1) When the channel buffer is empty, receiving will either:
+            (a) Block until an item is available (until the sending part
+            sends an item).
+            (b) Not block, and the received data will be zeros.
+        (2) When the channel buffer is not empty, receiving will either:
+            (a) Receive a single item, being the oldest item put in the
+            channel (FIFO).
+            (b) Receive all items available, accumulated.
+
+    Parameters
+    ----------
+    shape : tuple
+        Shape of the InPort of the Process.
+    buffer_size : optional, int
+        Buffer size of the extractor_channel.
+    extractor_channel_config : optional, ChannelConfig
+        Configuration object specifying how the src_port behaves when the
+        buffer is full and how the dst_port behaves when the buffer is empty
+        and not empty.
+    """
+    def __init__(self,
+                 shape: tuple[int, ...],
+                 buffer_size: ty.Optional[int] = 50,
+                 extractor_channel_config: ty.Optional[ChannelConfig] =
+                 ChannelConfig()) -> None:
+        super().__init__()
+
+        validate_shape(shape)
+        validate_buffer_size(buffer_size)
+        validate_channel_config(extractor_channel_config)
+
+        self._shape = shape
+
+        self._extractor_channel_config = extractor_channel_config
+
+        self._multi_processing = MultiProcessing()
+        self._multi_processing.start()
+
+        self._extractor_channel = \
+            PyPyChannel(message_infrastructure=self._multi_processing,
+                        src_name="src",
+                        dst_name="dst",
+                        shape=self._shape,
+                        dtype=float,
+                        size=buffer_size)
+        self._extractor_channel_dst_port = self._extractor_channel.dst_port
+        self._extractor_channel_dst_port.start()
```

Comment:
I'm not positive, but I don't think you should be calling start on the channel until after the runtime has been initialized.

Reply:
I also wanted to implement it this way. However, if we do that, and use the Injector or Extractor as we do in the `test_run_steps_non_blocking` tests (found in both `test_injector.py` and `test_extractor.py`) (which is how these Processes are typically expected to be used) will probably not work most of the times.
This is because the call to run(), when using the RunSteps(blocking=False) condition will not block, and by the time the next line executes (i.e we start sending or receiving), Runtime might not be initialized yet, thus an error would be thrown.
We could of course make sure to sleep a bit after the call to run() to give some time for the compilation to finish and for the Runtime to be initialized, but this amount of time needed for this could very much vary depending on the network being run...

Reply:
The runtime initialization should never be threaded/subprocessed. When you call proc.run, if you haven't already called proc.create_runtime, it will always block at least long enough to setup all the messaging and runtime infrastructure, create ports and channels, etc. The only thing that should be non-blocking is the work that happens inside the process builders and the run code itself.

Now, you might still be right that it could lead to errors if implemented that way, but those errors would be bugs that should be fixed, not just acceptable indeterminism.

Reply:
You are right. The call to run() with RunSteps(blocking=False) actually returns after the Runtime is initialized (it actually even returns after the Runtime starts running...), so the issue I was pointing to is not really an issue here.

That being said, starting the channel only after the Runtime is initialized requires us to be able to trigger some callback in the Process after the Runtime is initialized, which is not straightforward (see comment under init).